gateways/custom/storagegrid/start.go

```diff
@@ -81,204 +115,189 @@ func filterEvent(notification *storageGridNotification, sg *StorageGridEventConf
+		return
+	}
+	b, err := qson.ToJSON(parsedURL)
+	if err != nil {
+		rc.gatewayConfig.ErrChan <- err
+		return
+	}
 
-		// server with same port is already started by another configuration
-		if sg.Mux == nil {
-			mutex.Lock()
-			sg.Mux, ok = activeServers[sg.Port]
-			if !ok {
-				ce.startHttpServer(sg, config)
-			}
-			mutex.Unlock()
-		}
+	var notification *storageGridNotification
+	err = json.Unmarshal(b, &notification)
+	if err != nil {
+		rc.gatewayConfig.ErrChan <- err
+		return
+	}
 
-		sg.Mux.HandleFunc(sg.Endpoint, func(writer http.ResponseWriter, request *http.Request) {
-			ce.Log.Info().Str("endpoint", sg.Endpoint).Str("http-method", request.Method).Msg("received a request")
-			// Todo: find a better to handle route deletion
-			if _, ok := activeRoutes[sg.Port][sg.Endpoint]; ok {
-				body, err := ioutil.ReadAll(request.Body)
-				if err != nil {
-					ce.Log.Error().Err(err).Msg("failed to parse request body")
-					common.SendErrorResponse(writer)
-					config.Active = false
-					config.ErrChan <- err
-					return
-				}
+	rc.configExecutor.Log.Info().Str("config-key", rc.gatewayConfig.Data.Src).Interface("notification", notification).Msg("parsed notification")
+	if !filterEvent(notification, rc.sgConfig) || !filterName(notification, rc.sgConfig) {
+		rc.configExecutor.Log.Warn().Str("config-key", rc.gatewayConfig.Data.Src).Interface("notification", notification).
+			Msg("discarding notification since it did not pass all filters")
+	}
 
-				ce.Log.Info().Str("endpoint", sg.Endpoint).Str("http-method", request.Method).
-					Str("payload", string(body)).Msg("dispatching event to gateway-processor")
-
-				switch request.Method {
-				case http.MethodPost, http.MethodPut:
-					body, err := ioutil.ReadAll(request.Body)
-					if err != nil {
-						ce.Log.Error().Err(err).Str("config-key", config.Data.Src).Msg("failed to parse request body")
-					} else {
-						ce.Log.Info().Str("config-key", config.Data.Src).Str("msg", string(body)).Msg("msg body")
-					}
-
-				case http.MethodHead:
-					ce.Log.Info().Str("config-key", config.Data.Src).Str("method", http.MethodHead).Msg("received a request")
-					respBody = ""
-				}
-				writer.WriteHeader(http.StatusOK)
-				writer.Header().Add("Content-Type", "text/plain")
-				writer.Write([]byte(respBody))
+	rc.gatewayConfig.DataChan <- b
+}
 
-				// notification received from storage grid is url encoded.
-				parsedURL, err := url.QueryUnescape(string(body))
-				if err != nil {
-					config.Active = false
-					config.ErrChan <- err
-					return
-				}
-				b, err := qson.ToJSON(parsedURL)
-				if err != nil {
-					config.Active = false
-					config.ErrChan <- err
-					return
-				}
+// routeDeactivateHandler handles routes that are not active
+func (rc *routeConfig) routeDeactivateHandler(writer http.ResponseWriter, request *http.Request) {
+	rc.configExecutor.Log.Info().Str("endpoint", rc.sgConfig.Endpoint).Str("http-method", request.Method).Msg("route is not active")
```

Comment:
do you need to delete from `activeServers` here?

Reply:
nope, instead of deleting it from activeServers, the code assigns http handler that returns 404 for the route. This essentially makes route deactive.